x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/AuthorizationUtils.java

```diff
@@ -150,6 +197,163 @@ public static void switchUserBasedOnActionOriginAndExecute(
+        final Set<String> children = PRE_AUTHORIZED_CHILD_ACTIONS.get(parent);
+        return children != null && (parent.equals(child) || children.contains(child));
+    }
+
+    public static void maybePreAuthorizeChildAction(
+        ThreadContext threadContext,
+        ClusterState clusterState,
+        String childAction,
+        DiscoveryNode destinationNode,
+        TransportRequest request,
+        Version version
+    ) {
+
+        AuthorizationEngine.AuthorizationContext parentContext = extractAuthorizationContext(threadContext, childAction);
+        if (parentContext == null) {
+            return;
+        }
+
+        final IndicesAccessControl indicesAccessControl = parentContext.getIndicesAccessControl();
+        if (indicesAccessControl == null) {
+            // This can happen if the parent request was authorized by index name only - e.g. bulk request
+            // A missing IAC is not an error, but it means we can't safely tie authz of the child action to the parent authz
+            return;
+        }
+
+        // Just a sanity check. If we ended up here, the authz should have been granted.
+        if (indicesAccessControl.isGranted() == false) {
+            return;
+        }
+
+        final IndicesRequest indicesRequest;
+        if (request instanceof IndicesRequest) {
+            indicesRequest = (IndicesRequest) request;
+        } else {
+            // We can only pre-authorize indices request.
+            return;
+        }
+
+        final Role role = RBACEngine.maybeGetRBACEngineRole(parentContext.getAuthorizationInfo());
+        if (role == null) {
+            // If role is null, it means a custom authorization engine is in use, hence we cannot do the optimization here.
+            return;
+        }
+
+        // We can't safely pre-authorize actions if DLS or FLS is configured without passing IAC as well with the authorization result.
+        // For simplicity, we only pre-authorize actions when FLS and DLS are not configured, otherwise we would have to compute and send
+        // indices access control as well, which we want to avoid with this optimization.
+        if (role.hasFieldOrDocumentLevelSecurity()) {
+            return;
+        }
+
+        final String parentAction = parentContext.getAction();
+        if (shouldPreAuthorizeChildAction(parentAction, childAction) == false) {
+            return;
+        }
+
+        if (clusterState.nodes().getLocalNode().equals(destinationNode)) {
+            // Child actions targeting local (same) node are handled differently and already pre-authorized by parent action.
+            return;
+        }
+
+        if (clusterState.nodes().nodeExists(destinationNode) == false) {
+            // We can only pre-authorize actions targeting node which belongs to the same cluster.
+            // TODO: Ensure we remove any existing pre-authorization before sending request to remote cluster node.
+            return;
+        }
```

Comment:
Correct. In my case I only need to know "whether the node belongs to the local cluster". Discussion was around having a generic way to make this distinction to avoid the duplication of effort. I'm personally fine with both approaches, but maybe slightly preferring to have one way which benefits both - in this case to make a distinction if this is a connection to a node belonging the remote cluster.

Reply:
I have no issue with a single approach to answer both questions.